asyncio_mqtt/client.py

```diff
@@ -2,49 +2,90 @@
 import asyncio
 import logging
 import socket
+import ssl
 from contextlib import contextmanager, suppress
+from types import TracebackType
+from typing import (
+    Any,
+    AsyncGenerator,
+    AsyncIterator,
+    Awaitable,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Tuple,
+    Type,
+    cast,
+)
 
 try:
     from contextlib import asynccontextmanager
 except ImportError:
-    from async_generator import asynccontextmanager
-import paho.mqtt.client as mqtt
-from .error import MqttError, MqttCodeError, MqttConnectError
+    from async_generator import asynccontextmanager  # type: ignore
 
+import paho.mqtt.client as mqtt  # type: ignore
+
+from .error import MqttCodeError, MqttConnectError, MqttError
+from .types import (
+    PayloadType,
+    ProtocolType,
+    T,
+)
 
 MQTT_LOGGER = logging.getLogger("mqtt")
 MQTT_LOGGER.setLevel(logging.WARNING)
 
 
+# TODO: This should be a (frozen) dataclass (from Python 3.7)
+# when we drop Python 3.6 support
+class Will:
+    def __init__(
+        self,
+        topic: str,
+        payload: PayloadType = None,
+        qos: int = 0,
+        retain: bool = False,
+        properties: mqtt.Properties = None,
+    ):
+        self.topic = topic
+        self.payload = payload
+        self.qos = qos
+        self.retain = retain
+        self.properties = properties
+
+
 class Client:
     def __init__(
         self,
-        hostname,
-        port=1883,
+        hostname: str,
+        port: int = 1883,
         *,
-        username=None,
-        password=None,
-        logger=None,
-        client_id=None,
-        tls_context=None,
-        protocol=None,
-        will=None,
-        clean_session=None,
-        transport="tcp",
+        username: Optional[str] = None,
+        password: Optional[str] = None,
+        logger: logging.Logger = MQTT_LOGGER,
```

Comment:
Just note that this is a breaking change since the user can no longer explicitly use `logger=None`. I'm not opposed to the change as such. We just have to remember this due to semantic versioning.

Reply:
Ah, I did wonder why that was `None`. I'll revert that.

Reply:
OK no probs, I'll revert that.